### PR TITLE
[render-test] Add support for collecting network metrics

### PR DESCRIPTION
**Pull request by jmalanen**
Fixes mapbox/mapbox-gl-native-team/issues/43

### CODE REVIEW

File: render-test/runner.cpp [hunk 1]
```diff
@@ -783,6 +832,8 @@ TestRunner::Impl::Impl(const TestMetadata& metadata)
 bool TestRunner::run(TestMetadata& metadata) {
     AllocationIndex::setActive(false);
     AllocationIndex::reset();
+    ProxyFileSource::setTrackingActive(false);
```
Comment (tmpsantos) [inline render-test/runner.cpp]: Nit: To be consistent with `AllocationIndex` can we rename this to `setActive` and `reset`?

File: render-test/runner.cpp [hunk 2]
```diff
@@ -294,6 +295,31 @@ bool TestRunner::checkRenderTestResults(mbgl::PremultipliedImage&& actualImage,
...
     }
 
+    // Check network metrics.
+    for (const auto& expected : metadata.expectedMetrics.network) {
+        auto actual = metadata.metrics.network.find(expected.first);
+        if (actual == metadata.metrics.network.end()) {
+            metadata.errorMessage = "Failed to find network probe: " + expected.first;
+            return false;
+        }
+        if (actual->second.requests > expected.second.requests) {
```
Comment (tmpsantos) [inline render-test/runner.cpp]: `!=`? Making less requests should also be an error. For instance a regression could make us request less fonts.

File: render-test/runner.cpp [hunk 3]
```diff
@@ -294,6 +295,31 @@ bool TestRunner::checkRenderTestResults(mbgl::PremultipliedImage&& actualImage,
...
+            std::stringstream ss;
+            ss << "Number of requests at probe \"" << expected.first << "\" is " << actual->second.requests
+               << ", expected is " << expected.second.requests << ".";
+
+            metadata.errorMessage = ss.str();
+            return false;
+        }
+        if (actual->second.transferred > expected.second.transferred) {
```
Comment (tmpsantos) [inline render-test/runner.cpp]: Ditto. We will probably need to add tolerance if we make this `!=`.

File: render-test/file_source.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    active = active_;
+}
+
+// static
+bool ProxyFileSource::isTrackingActive() {
+    return active;
+}
+
+// static
+void ProxyFileSource::resetMetrics() {
```
Comment (tmpsantos) [inline render-test/file_source.cpp]: This method is not needed, could be implicit inside `::setTrackingActive` if you change the state of `active`.

File: render-test/tests/should-fail.json [hunk 5]
```diff
@@ -2,5 +2,7 @@
   "tests/file-size/fail-size-is-over": "Should fail, size is bigger than expected.",
   "tests/file-size/fail-size-is-under": "Should fail, size is smaller than expected.",
   "tests/file-size/fail-file-doesnt-match": "Should fail, doesn't match the expectation.",
-  "tests/file-size/fail-file-not-found": "Should fail, file not found."
+  "tests/file-size/fail-file-not-found": "Should fail, file not found.",
+  "tests/network/fail-requests": "Should fail, number of requests higher than expected.",
+  "tests/network/fail-transferred": "Should fail, amount of transferred data higher than expected."
```
Comment (tmpsantos) [inline render-test/tests/should-fail.json]: Maybe add a test for when count and transferred is greater/lower.

File: render-test/file_source.cpp (hunk 4 above)
Comment (jmalanen) [reply render-test/file_source.cpp]: Done

File: render-test/tests/should-fail.json (hunk 5 above)
Comment (jmalanen) [reply render-test/tests/should-fail.json]: Done

File: render-test/runner.cpp (hunk 3 above)
Comment (jmalanen) [reply render-test/runner.cpp]: Done

File: render-test/runner.cpp (hunk 2 above)
Comment (jmalanen) [reply render-test/runner.cpp]: Done

**Pull request merged**